### ISSUE TITLE
PGPLiteralDataGenerator: Text mode does not convert line endings to <CR><LF>

### ISSUE THREAD

**Issue by vanitasvitae** [bcgit/bc-java#1035]
As pointed out by @bjansen [here](https://github.com/pgpainless/pgpainless/discussions/191), Bouncycastle does not convert line endings of data inside a Literal Data packet to <CR><LF>, even though the format is set to `t` (text mode).

I'm not sure, how easy it would be to fix this issue. Maybe it is enough to wrap the `WrappedGeneratorStream` into a `FilteredOutputStream` which replaces the line endings, before returning the stream in `PGPLiteralDataGenerator.open()`?

**Comment by vanitasvitae** [comment]
I did some experiments in https://github.com/pgpainless/bc-java/commit/470fb6068b94d7f13fb75ed244599345b6364d89

I'm not yet convinced that this is the easy fix, since some tests are breaking if I apply my patch.

**Comment by tonywasher** [comment]
The rfc4660 Spec states that in text mode the signature is calculated over the literal data with its line ends converted to <CR><LF>. It emphatically does NOT say that the actual text is to be converted.

If you think about it, the clear intent is that if a signature is calculated on a text file on linux (with LF line endings) and the text file is transferred via FTP to a windows machine such that it now has CRLF line endings, the signature can also be transferred  and verified against the transferred file.

The original report refers to a problem calculating the signature between pgpainless (on top of bouncyCastle) and  an IBM PGP implementation on the mainframe. It should be noted that mainframe text files are organised as a sequence of records and as such there is no such thing as a line ending on the mainframe!!!

I suspect that the problem is firmly in the IBM PGP Mainframe implementation. In particular you should test the interoperability between GnuPG (which should be viewed as the canonical implementation) and the IBM PGP implementation. I am certain that you will find that BouncyCastle matches GnuPG and that it is the mainframe implementation that is out of step.

**Comment by bjansen** [comment]
@tonywasher perhaps you are thinking of *detached signatures*, for which line endings are canonicalized to compute the signature but not in the text file?

**Comment by tonywasher** [comment]
Yes you are correct. I was reading the section where the signature is defined rather than the Literal data section., 

I expect that the mainframe end assumes that the data that it receives in the literal data packet has CRLF endings and hence does not play funny games with the line endings.

I think you are on the right lines with your patch, except that you patch will only work on Unix systems, The native line endings are CRLF on windows LF on unix and CR on mac,. The actual native line ending is held in the system property line.separator and the wrapper code should convert to and from CRLF <-> line.separator

**Comment by vanitasvitae** [comment]
> The actual native line ending is held in the system property line.separator and the wrapper code should convert to and from CRLF <-> line.separator

Good point!

**Comment by bjansen** [comment]
OK, here's another interesting experiment: I generated an encrypted file using BouncyGPG/BouncyCastle and tried to decrypt it with `gpg`.

When my code converts line endings to `\r\n`, `gpg` is happy. But when I skip the conversion, `gpg` starts complaining that the signature is invalid:

```
gpg: encrypted with rsa4096 key, ID 3FE643F19DA9937A, created 2020-07-16
      "key1 <key1@example.com>"
gpg: encrypted with rsa4096 key, ID 23BFB3A0E641BE6D, created 2020-07-16
      "key2 <key2@example.com>"
hello
worldgpg: Signature made Tue Oct  5 16:45:57 2021 CEST
gpg:                using RSA key 6A839BD37853583D
gpg:                issuer "key2@example.com"
gpg: BAD signature from "key2 <key2@example.com>" [unknown]
```

**Comment by vanitasvitae** [comment]
@bjansen could you share armored test key and messages for reference?
A handy tool for this is https://dump.sequoia-pgp.org/ which allows you to create permalinks.

**Comment by tonywasher** [comment]
@vanitasvitae a further point on your patch. You have only intercepted the **int read()** call. You will also need to intercept the **int read(byte[] b,  int off, int len)** call and point it through your patched **int read()**. Apparently FilterOutputStream already does this.

**Comment by vanitasvitae** [comment]
Yeah, that patch needs some more work before its ready to be opened as a PR.
Thanks for the feedback in any case :)
